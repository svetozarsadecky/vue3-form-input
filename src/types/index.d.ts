declare module 'vue3-form-input';
