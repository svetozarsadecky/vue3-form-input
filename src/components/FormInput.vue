<script lang="ts">
export default {
    name: 'FormInput',
};
</script>

<template>
    <div :class="rootClasses">
        <div :class="elementClassName.field">
            <div
                v-if="hasSlotPrepend"
                :class="elementClassName.prepend"
                :data-test="elementClassName.prepend"
            >
                <slot
                    name="prepend"
                ></slot>
            </div>

            <label
                :for="input.id"
                :class="elementClassName.wrapper"
            >
                <span
                    v-if="showHider"
                    :class="elementClassName.hider"
                ></span>
                <textarea
                    v-if="isTextArea"
                    :id="input.id"
                    ref="refInput"
                    v-model="localValue"
                    :autocomplete="autocomplete"
                    :class="inputClassName"
                    :name="input.name"
                    :placeholder="translate(input.placeholder)"
                    :readonly="isReadonly"
                    :required="isRequired"
                    :disabled="isDisabled"
                    :autofocus="input.autofocus"
                    :rows="input.rows"
                    v-bind="bindAttrsInput"
                    data-test="form-input__input"
                    @blur="onBlur"
                    @focus="onFocus"
                    @input="onInput"
                    @keydown="onKeydown"
                    @keyup="onKeyup"
                ></textarea>

                <input
                    v-else
                    :id="input.id"
                    ref="refInput"
                    v-model="localValue"
                    :accept="input.accept"
                    :autocomplete="autocomplete"
                    :class="inputClassName"
                    :multiple="input.multiple"
                    :name="input.name"
                    :pattern="input.pattern"
                    :step="input.step"
                    :min="input.min"
                    :max="input.max"
                    :placeholder="translate(input.placeholder)"
                    :readonly="isReadonly"
                    :disabled="isDisabled"
                    :required="isRequired"
                    :autofocus="input.autofocus"
                    :type="type"
                    data-test="form-input__input"
                    v-bind="bindAttrsInput"
                    @blur="onBlur"
                    @focus="onFocus"
                    @input="onInput"
                    @keydown="onKeydown"
                    @keyup="onKeyup"
                >

                <span
                    v-if="hasSlotLabel"
                    class=""
                    :class="elementClassName.label"
                    :data-test="elementClassName.label"
                >
                    <slot name="label">
                        {{ translate(input.label) }}
                    </slot>
                </span>
            </label>

            <div
                v-if="hasSlotSuffix"
                :class="elementClassName.suffix"
                :data-test="elementClassName.suffix"
            >
                <slot
                    name="suffix"
                    :is-state-visible="showState"
                ></slot>
            </div>

            <div
                v-if="hasSlotAppend"
                :class="elementClassName.append"
                :data-test="elementClassName.append"
            >
                <slot
                    name="append"
                    :errors="errors"
                >
                </slot>
            </div>
        </div>

        <div
            v-if="hasSlotMessage"
            class="form-input__message text-secondary mt-1"
            :class="elementClassName.message"
            :data-test="elementClassName.message"
        >
            <slot name="message"></slot>
        </div>

        <div v-if="showState">
            <FormErrors
                :errors="errors"
                data-test="form-input__errors"
            ></FormErrors>
        </div>
    </div>
</template>

<script lang="ts" setup>
import {
    ref,
    withDefaults,
    computed,
    watch,
    onMounted,
    onBeforeUnmount,
    useSlots,
    reactive,
} from 'vue';
import { INPUT_TYPES } from '@/constants/form';
import { generateElementClassNames, getClassName } from '@/helpers';
import FormErrors from '@/components/FormErrors.vue';
import type {
    IInput,
    TFormInputValue,
} from '@/types';
import FormItem from '@/mixins/FormItem';
import FormItemValidator from '@/mixins/FormItemValidator';
import FormItemInstaller from '@/mixins/FormItemInstaller';


// TODO: Add alternative non floating prop for input.
//  So label can coexist with placeholder
interface Props {
    className?: string | null
    bindAttrsInput?: Record<string, IInput> | null
    groupName?: string
    input: IInput
    formErrors?: Array<string>
    modelValue?: TFormInputValue
    inputClass?: string
    // TODO: Needs maybe more specific typing
    trans?: ((...args: any) => string)
    type?: INPUT_TYPES
}

interface Emits {
    (e: 'blur', event: Event): void
    (e: 'focus', event: Event): void
    (e: 'eager', event: Event): void
    (e: 'keydown', event: Event): void
    (e: 'keyup', event: Event): void
    (e: 'update:modelValue', value: TFormInputValue): void
}

/**
 * Reactive Data
 */
const emit = defineEmits<Emits>();
const props = withDefaults(defineProps<Props>(), {
    className: null,
    groupName: '',
    formErrors: () => ([]),
    modelValue: '',
    inputClass: '',
    type: undefined,
    trans: undefined,
    bindAttrsInput: null,
});
const {
    label,
    message,
    append,
    suffix,
    prepend,
} = useSlots();
const localValue = ref<TFormInputValue>(
    !props.modelValue
        ? ''
        : props.modelValue
);
const refInput = ref<HTMLInputElement | null>(null);
const clearCalled = ref<boolean>(false);

/**
 * Computed
 */
const isFilled = computed<boolean>(() => `${localValue.value}`.length > 0);
const isTextArea = computed<boolean>(() => props.input.type === INPUT_TYPES.TEXTAREA);
const showHider = computed(() => isTextArea.value && props.input.label);
const autocomplete = computed(() => props.input.autocomplete || 'off');

// Computed Slots
const hasSlotLabel = computed<boolean>(() => !!props.input.label || (!!label && label().length > 0));
const hasSlotMessage = computed<boolean>(() => !!message && message().length > 0);
const hasSlotSuffix = computed<boolean>(() => !!suffix && suffix().length > 0);
const hasSlotAppend = computed<boolean>(() => !!append && append().length > 0);
const hasSlotPrepend = computed<boolean>(() => !!prepend && prepend().length > 0);

// Extend with FormItem general settings
const {
    className,
    inputClassName,

    type,

    isHidden,
    isReadonly,
    isDisabled,
    isRequired,
} = FormItem(
    localValue,
    props,
    'form-input',
);

// Extend with FormItemInstaller to add lifecycle hooks
const {
    onItemMounted,
    onItemBeforeUnmount,
} = FormItemInstaller();

// Extend with FormItemValidator to add validator logic
const {
    Validator,
    errors,
    showState,
    hasErrors,
    validate,
    setState,
} = FormItemValidator(
    props,
    localValue,
    refInput
);

// Computed classes
const rootClasses = computed<Record<string, boolean>>(() => ({
    [getClassName(className.value)]: true,
    [getClassName(className.value, null, 'is-filled')]: isFilled.value,
    [getClassName(className.value, null, 'has-error')]: hasErrors.value,
    [getClassName(className.value, null, 'has-no-label')]: !hasSlotLabel.value,
    [getClassName(className.value, null, 'is-required')]: isRequired.value,
    [getClassName(className.value, null, 'is-readonly')]: isReadonly.value,
    [getClassName(className.value, null, 'is-hidden')]: isHidden.value,
    [getClassName(className.value, null, 'is-disabled')]: isDisabled.value,
    [getClassName(className.value, null, 'is-valid')]: showState.value && !hasErrors.value,
    [getClassName(className.value, null, 'textarea')]: isTextArea.value,
}));
const elementClassName = reactive(generateElementClassNames(
    className.value,
    [
        'append',
        'prepend',
        'suffix',
        'field',
        'hider',
        'label',
        'wrapper',
        'message',
    ]
));

/**
 * Watchers
 */
watch(
    () => props.modelValue,
    (value) => {
        localValue.value = !value
            ? ''
            : value;

        /**
         * When value change hide errors.
         * and validate by validator type
         * @type {boolean}
         */
        if (clearCalled.value) {
            return;
        }

        Validator.validateByEventType(localValue.value, 'input', setState);
        clearCalled.value = false;
    }
);

watch(
    () => props.formErrors,
    (value: Record<string, string> | Array<string>) => {
        if (value && Object.keys(value.length > 0)) {
            setState();
            errors.value = props.formErrors;
        }
    },
);

// Events
function onFocus(event: Event): void {
    emit('focus', event);
}

function onBlur(event: Event): void {
    const target = event.target as HTMLInputElement | HTMLTextAreaElement;

    if (target && ('value' in target)) {
        target.value = String(localValue.value);
    }

    Validator.validateByEventType(localValue.value, event.type, setState);
    emit('blur', event);
}

function onInput(event: Event): void {
    const target = event.target as HTMLInputElement | HTMLTextAreaElement;

    Validator.validateByEventType(localValue.value, event.type, setState);
    emit('eager', event);

    if (target && ('value' in target)) {
        emit('update:modelValue', target.value);
    }
}

function onKeydown(event: Event): void {
    emit('keydown', event);
}

function onKeyup(event: Event): void {
    emit('keyup', event);
}

// Methods
function translate(key: string): string {
    if (typeof props.trans === 'function') {
        return props.trans(key);
    }

    return key;
}

function clear(): void {
    clearCalled.value = true;
    localValue.value = '';

    emit('update:modelValue', localValue.value);

    errors.value = [];
    showState.value = false;
}


/**
 * Lifecycle hooks subscribing/unsubscribing from active Form items
 */
onMounted(() => {
    onItemMounted();
});

onBeforeUnmount(() => {
    onItemBeforeUnmount();
});

defineExpose({
    clear,
    groupName: props.groupName,
    formErrors: errors,
    validate,
});
</script>

<style lang="scss">
@import "@/scss/form-input";
</style>
