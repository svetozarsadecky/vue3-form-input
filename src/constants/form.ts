export enum INPUT_TYPES {
    TEXT = 'text',
    TEXTAREA = 'textarea',
    NUMBER = 'number',
    PASSWORD = 'password',
    EMAIL = 'email',
    HIDDEN = 'hidden',
}
