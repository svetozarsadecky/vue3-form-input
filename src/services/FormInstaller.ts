import { ref, shallowRef } from 'vue';
import type { App, ComponentInternalInstance } from 'vue';
import type { Emitter } from 'mitt';
import type { IMxFormHandler, TEvents } from '@/types/index';
import mitt from 'mitt';

interface IFormInstallerOptions {
    event: Emitter<TEvents>
}

enum FUNCTION_NAMES {
    VALIDATE = 'validate',
    CLEAR = 'clear',
}

// TODO: Add integration tests
function install(app: App, options: IFormInstallerOptions): void {
    const isCorrectVersion = formInstallerCheckVersion(app);

    if (!isCorrectVersion) {
        return;
    }

    if (app.config.globalProperties.$formHandler) {
        return;
    }

    const event: Emitter<TEvents> = mitt();
    const formHandler: IMxFormHandler = {
        activeItems: shallowRef([]),
        errors: ref({}),
        /**
         * Global validation method
         * used together with active subscribed form items
         *
         * Can call default validate method on component
         * if he is part of chosen form group
         * or even call a method with custom name if necessary
         *
         * @param groupName     - form group name
         * @param functionName  - custom validation method name
         */
        validate(groupName, functionName) {
            formInstallerCallFunctionOnItem.call(
                this,
                groupName,
                functionName || FUNCTION_NAMES.VALIDATE
            );
            formInstallerSetGatherErrors.call(this, groupName);

            return !this.hasErrors(groupName);
        },
        /**
         * Clears validation errors on form items
         * belonging to a chosen group
         *
         * @param groupName - form group name
         */
        clear(groupName: string) {
            formInstallerCallFunctionOnItem
                .call(this, groupName, FUNCTION_NAMES.CLEAR);
        },
        /**
         * Returns boolean value for specific form group's error state
         * @param groupName
         */
        hasErrors(groupName: string) {
            return this.errors.value[groupName].length > 0;
        },
        /**
         * Returns all errors in array for specific form group
         * @param groupName
         */
        getErrors(groupName) {
            formInstallerSetGatherErrors.call(this, groupName);

            return this.errors.value[groupName];
        },
    };

    app.config.globalProperties.$formHandler = formHandler;
    app.provide('formHandler', formHandler);
    app.provide('formHandlerEvent', event);

    event.on('subscribe', formInstallerSubscribeItem.bind(formHandler));
    event.on('unsubscribe', formInstallerUnsubscribeItem.bind(formHandler));
}

/**
 * Gather and set errors from all the form items
 * with specific group name to be accessible on the formHandler instance
 *
 * @param groupName
 */
function formInstallerSetGatherErrors(this: IMxFormHandler, groupName: string): void {
    this.errors.value[groupName] = [];

    formInstallerGetActiveGroupItems.call(this, groupName).forEach(
        async (item) => {
            const exposed = item.exposed;

            if (!exposed || exposed.formErrors.value.length === 0) {
                return;
            }

            this.errors.value[groupName].push(...exposed.formErrors.value);
        }
    );


}

/**
 * Checks if current Vue version is 3 or higher
 * @param app
 */
function formInstallerCheckVersion(app: App): boolean {
    const version = Number(app.version.split('.')[0]);

    if (version < 3) {
        console.warn('This plugin requires Vue 3');

        return false;
    }

    return true;
}

/**
 * Add form item to active items array
 * on mounted hook
 * @param item - form item component (input, select...)
 */
function formInstallerSubscribeItem(
    this: IMxFormHandler,
    item: ComponentInternalInstance | null
): void {
    if (item) {
        this.activeItems.value.push(item);
    }
}

/**
 * Remove form item from active items array
 * on before destroy hook
 * @param item - form item component (input, select...)
 */
function formInstallerUnsubscribeItem(
    this: IMxFormHandler,
    item: ComponentInternalInstance | null
): void {
    if (!item) {
        return;
    }

    const index = this.activeItems.value.indexOf(item);

    if (index !== -1) {
        this.activeItems.value.splice(index, 1);
    }
}

/**
 * Calls function on group of form items
 * if function exists and is exposed by form item
 * @param groupName          - group name for group of form items
 * @param functionName       - name of function to call
 */
function formInstallerCallFunctionOnItem(
    this: IMxFormHandler,
    groupName: string,
    functionName: FUNCTION_NAMES | string
): void {
    formInstallerGetActiveGroupItems.call(this, groupName)
        .forEach((item) => {
            const exposed = item.exposed;

            if (exposed && exposed[functionName]) {
                exposed[functionName]();
            }
        });
}

/**
 * Returns all active form items that are part of one group
 * @param groupName
 */
function formInstallerGetActiveGroupItems(
    this: IMxFormHandler,
    groupName: string,
): Array<ComponentInternalInstance> {
    return this.activeItems.value.filter((item: ComponentInternalInstance) =>
        !!item?.exposed?.groupName
        && item.exposed.groupName === groupName);
}

const plugin = {
    install,
};

export default plugin;
