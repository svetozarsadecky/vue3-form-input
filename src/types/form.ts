import type { INPUT_TYPES } from '@/constants/form';
import type { ComponentInternalInstance, Ref, ShallowRef } from 'vue';

export enum INPUT_EVENT_TYPES {
    BLUR_INPUT = 'onBlurThenOnInput',
    INPUT = 'onInput',
    BLUR = 'onBlur',
}

export type TFormInputValue = string | number;

export type TTransProxy = (value: string, ...args: never) => string;

export interface IFormInputValidator {
    validator: string | ((value: TFormInputValue, ...args: never) => boolean)
    message?: string
}

export interface IFormValidatorOptions {
    validators?: Array<IFormInputValidator>,
    event?: INPUT_EVENT_TYPES,
}

export interface IInputTextarea {
    autocomplete?: string
    autofocus?: boolean
    disabled?: boolean
    id?: string
    name?: string
    placeholder?: string
    readonly?: boolean
    required?: boolean
    rows?: number | string
    value?: string
    type?: INPUT_TYPES
    validatorEvent?: string
    validators?: Array<IFormInputValidator>
}

export interface IInputCheckbox {
    name?: string
    id?: string
    value: string | boolean | null
    readonly?: boolean
    label?: string
    disabled?: boolean
    required?: boolean
    trueValue?: string
    falseValue?: string
    autocomplete?: string
    validatorEvent?: string
    validators?: Array<IFormInputValidator>
}

export interface IInput extends IInputTextarea {
    label?: string
    validatorEvent?: string
    accept?: string
    multiple?: boolean
    pattern?: string
    step?: number | string
    max?: number | string
    min?: number | string
}

export interface IMxFormHandler {
    activeItems: ShallowRef<Array<ComponentInternalInstance>>
    errors: Ref<Record<string, Array<string>>>
    clear: (groupName: string) => void,
    validate: (groupName: string, functionName?: string) => boolean,
    hasErrors: (groupName: string) => boolean,
    getErrors: (groupName: string) => Array<string>,
}
