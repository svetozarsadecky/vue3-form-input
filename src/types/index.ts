import type { ComponentInternalInstance } from 'vue';

export * from './form';

export type TEvents = {
    subscribe: ComponentInternalInstance | null,
    unsubscribe: ComponentInternalInstance | null,
}

/**
 * Generic type for ***`Input`*** elements which contains details of target element
 */
export interface IExtendedInputEvent extends InputEvent {
    target: HTMLInputElement;
}

