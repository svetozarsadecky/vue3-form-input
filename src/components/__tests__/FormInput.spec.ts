import { describe, it, expect, vi } from 'vitest';
import { DOMWrapper, mount } from '@vue/test-utils';
import FormInput from '@/components/FormInput.vue';
import { reactive } from 'vue';
import { FORM_INPUT_FULL } from '@/components/__tests__/mocks/formInput';
import type { IInput } from '@/types/form';
import FormInstaller from '@/services/FormInstaller';
import { INPUT_TYPES } from '@/constants/form';
import { INPUT_EVENT_TYPES } from '@/types/form';

const INPUT_SELECTOR = '[data-test=form-input__input]';
const STATE_SELECTOR = '[data-test=form-input__errors]';
const SLOT_PREPEND_SELECTOR = '[data-test=form-input__prepend]';
const SLOT_LABEL_SELECTOR = '[data-test=form-input__label]';
const SLOT_SUFFIX_SELECTOR = '[data-test=form-input__suffix]';
const SLOT_APPEND_SELECTOR = '[data-test=form-input__append]';
const SLOT_MESSAGE_SELECTOR = '[data-test=form-input__message]';

const input = reactive<IInput>(FORM_INPUT_FULL);

const defaultOptions = {
    floatingVue: {},
    // Options: {
    //      trans
    //      ...
    // }
    formInstaller: {},
};

vi.mock('@/mixins/FormItemInstaller', async () => {
    return {
        default: () => ({
            onItemMounted: vi.fn(),
            onItemBeforeUnmount: vi.fn(),
        })
    };
});

const factory = ({ props = {}, slots = {}, global = {} } = {}) => mount(FormInput, {
    props: {
        input,
        ...props,
    },
    slots: {
        ...slots,
    },
    global: {
        directives: {
            tooltip: vi.fn()
        },
        plugins: [[FormInstaller, { ...defaultOptions }]],
        ...global,
    }
});

describe('FormInput props, slots & events', () => {
    it('renders component', () => {
        const wrapper = factory();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('renders with input.required set to true', async () => {
        const wrapper = factory({
            props: {
                input: { ...input, required: true },
            }
        });

        expect(wrapper.attributes().class).toMatchInlineSnapshot('"form-input form-input--is-required"');
    });

    it('renders with input.disabled set to true', () => {
        const wrapper = factory({
            props: {
                input: { ...input, disabled: true },
            }
        });

        const inputEl: DOMWrapper<HTMLInputElement | HTMLTextAreaElement> = wrapper.find(INPUT_SELECTOR);

        expect(inputEl.element.disabled).toBe(true);
        expect(wrapper.attributes().class).toMatchInlineSnapshot('"form-input form-input--is-disabled"');
    });

    it('renders with input.readonly set to true', async () => {
        const wrapper = factory({
            props: {
                input: { ...input, readonly: true },
            }
        });

        expect(wrapper.attributes().class).toMatchInlineSnapshot('"form-input form-input--is-readonly"');
    });

    it('renders input type hidden', async () => {
        const wrapper = factory({
            props: {
                input: { ...input, type: INPUT_TYPES.HIDDEN },
            }
        });

        const inputEl: DOMWrapper<HTMLInputElement | HTMLTextAreaElement> = wrapper.find(INPUT_SELECTOR);

        expect(inputEl.element.type).toBe(INPUT_TYPES.HIDDEN);
        expect(wrapper.attributes().class).toMatchInlineSnapshot('"form-input form-input--is-hidden"');
    });

    it('renders with filled class when modelValue is set', async () => {
        const wrapper = factory({
            props: {
                modelValue: 'text',
            }
        });

        const inputEl: DOMWrapper<HTMLInputElement | HTMLTextAreaElement> = wrapper.find(INPUT_SELECTOR);

        expect(inputEl.element.value).toBe('text');
        expect(wrapper.attributes().class).toMatchInlineSnapshot('"form-input form-input--is-filled"');
    });

    it('renders textarea instead input', async () => {
        const wrapper = factory({
            props: {
                input: {
                    ...input,
                    type: INPUT_TYPES.TEXTAREA,
                    rows: 1,
                },
            }
        });

        const inputEl: DOMWrapper<HTMLInputElement | HTMLTextAreaElement> = wrapper.find(INPUT_SELECTOR);

        expect(inputEl.element.type).toBe(INPUT_TYPES.TEXTAREA);
        expect(wrapper.attributes().class).toMatchInlineSnapshot('"form-input form-input--textarea"');
    });

    it('renders slot content for prepend', async () => {
        const wrapper = factory({
            slots: {
                prepend: 'Some content',
            }
        });

        const slot: DOMWrapper<HTMLDivElement> = wrapper.find(SLOT_PREPEND_SELECTOR);

        expect(slot.exists()).toBe(true);
        expect(slot.html()).toContain('Some content');
    });

    it('renders slot content for label', async () => {
        const wrapper = factory({
            slots: {
                label: 'Some content',
            }
        });

        const slot: DOMWrapper<HTMLDivElement> = wrapper.find(SLOT_LABEL_SELECTOR);

        expect(slot.exists()).toBe(true);
        expect(slot.html()).toContain('Some content');
    });

    it('renders slot content for suffix', async () => {
        const wrapper = factory({
            slots: {
                suffix: 'Some content',
            }
        });

        const slot: DOMWrapper<HTMLDivElement> = wrapper.find(SLOT_SUFFIX_SELECTOR);

        expect(slot.exists()).toBe(true);
        expect(slot.html()).toContain('Some content');
    });

    it('renders slot content for append', async () => {
        const wrapper = factory({
            slots: {
                append: 'Some content',
            }
        });

        const slot: DOMWrapper<HTMLDivElement> = wrapper.find(SLOT_APPEND_SELECTOR);

        expect(slot.exists()).toBe(true);
        expect(slot.html()).toContain('Some content');
    });

    it('renders slot content for message', async () => {
        const wrapper = factory({
            slots: {
                message: 'Some content',
            }
        });

        const slot: DOMWrapper<HTMLDivElement> = wrapper.find(SLOT_MESSAGE_SELECTOR);

        expect(slot.exists()).toBe(true);
        expect(slot.html()).toContain('Some content');
    });

    it('emits update event on input', async () => {
        const wrapper = factory();
        await wrapper.setProps({
            modelValue: 'Some content',
        });

        const inputEl: DOMWrapper<HTMLInputElement> = wrapper.find(INPUT_SELECTOR);
        await inputEl.trigger('input');

        expect(inputEl.element.value).toMatchInlineSnapshot('"Some content"');
        expect(wrapper.emitted()).toMatchInlineSnapshot(`
          {
            "eager": [
              [
                Event {
                  "isTrusted": false,
                },
              ],
            ],
            "input": [
              [
                Event {
                  "isTrusted": false,
                },
              ],
            ],
            "update:modelValue": [
              [
                "Some content",
              ],
            ],
          }
        `);
    });

    it('emits focus event on input focus', async () => {
        const wrapper = factory();

        const inputEl: DOMWrapper<HTMLInputElement> = wrapper.find(INPUT_SELECTOR);
        await inputEl.trigger('focus');

        expect(wrapper.emitted()).toMatchInlineSnapshot(`
          {
            "focus": [
              [
                FocusEvent {
                  "isTrusted": false,
                },
              ],
            ],
          }
        `);
    });

    it('emits blur event on input blur', async () => {
        const wrapper = factory();

        const inputEl: DOMWrapper<HTMLInputElement> = wrapper.find(INPUT_SELECTOR);
        await inputEl.trigger('blur');

        expect(wrapper.emitted()).toMatchInlineSnapshot(`
          {
            "blur": [
              [
                FocusEvent {
                  "isTrusted": false,
                },
              ],
            ],
          }
        `);
    });

    it('emits keydown event on input keydown', async () => {
        const wrapper = factory();

        const inputEl: DOMWrapper<HTMLInputElement> = wrapper.find(INPUT_SELECTOR);
        await inputEl.trigger('keydown');

        expect(wrapper.emitted()).toMatchInlineSnapshot(`
          {
            "keydown": [
              [
                KeyboardEvent {
                  "isTrusted": false,
                },
              ],
            ],
          }
        `);
    });

    it('emits keyup event on input keyup', async () => {
        const wrapper = factory();

        const inputEl: DOMWrapper<HTMLInputElement> = wrapper.find(INPUT_SELECTOR);
        await inputEl.trigger('keyup');

        expect(wrapper.emitted()).toMatchInlineSnapshot(`
          {
            "keyup": [
              [
                KeyboardEvent {
                  "isTrusted": false,
                },
              ],
            ],
          }
        `);
    });

    it('renders text processed with trans function', async () => {
        const wrapper = factory({
            props: {
                trans(value: string) {
                    return `Processed ${value}`;
                },
                input: { ...input, label: 'My label' },
            }
        });

        const labelSlot = wrapper.find(SLOT_LABEL_SELECTOR);
        expect(labelSlot.html()).toContain('Processed My label');
    });
});

describe('FormInput errors and validation', () => {
    it('renders error state component on error', async () => {
        const wrapper = factory();
        const inputEl: DOMWrapper<HTMLInputElement> = wrapper.find(INPUT_SELECTOR);

        await inputEl.trigger('blur');

        const stateEl: DOMWrapper<HTMLInputElement> = wrapper.find(STATE_SELECTOR);
        expect(stateEl.exists()).toBe(true);
    });

    it('renders with error class on blur', async () => {
        const wrapper = factory();

        const inputEl: DOMWrapper<HTMLInputElement> = wrapper.find(INPUT_SELECTOR);
        await inputEl.trigger('blur');

        expect(wrapper.emitted()).toMatchInlineSnapshot(`
          {
            "blur": [
              [
                FocusEvent {
                  "isTrusted": false,
                },
              ],
            ],
          }
        `);
        expect(wrapper.attributes().class).toContain('form-input--has-error');
    });

    it('renders with error class on input', async () => {
        const wrapper = factory({
            props: {
                input: { ...input, validatorEvent: INPUT_EVENT_TYPES.INPUT },
            }
        });

        const inputEl: DOMWrapper<HTMLInputElement> = wrapper.find(INPUT_SELECTOR);
        await inputEl.trigger('input');

        expect(wrapper.emitted()).toMatchInlineSnapshot(`
          {
            "eager": [
              [
                Event {
                  "isTrusted": false,
                },
              ],
            ],
            "input": [
              [
                Event {
                  "isTrusted": false,
                },
              ],
            ],
            "update:modelValue": [
              [
                "",
              ],
            ],
          }
        `);
        expect(wrapper.attributes().class).toContain('form-input--has-error');
    });

    it('renders with error class on input then on blur', async () => {
        const wrapper = factory({
            props: {
                input: { ...input, validatorEvent: INPUT_EVENT_TYPES.BLUR_INPUT },
            }
        });

        const inputEl: DOMWrapper<HTMLInputElement> = wrapper.find(INPUT_SELECTOR);
        await inputEl.trigger('input');

        expect(wrapper.emitted()).toMatchInlineSnapshot(`
          {
            "eager": [
              [
                Event {
                  "isTrusted": false,
                },
              ],
            ],
            "input": [
              [
                Event {
                  "isTrusted": false,
                },
              ],
            ],
            "update:modelValue": [
              [
                "",
              ],
            ],
          }
        `);
        expect(wrapper.attributes().class).toContain('form-input--has-error');

        await wrapper.setProps({
            modelValue: 'wrong email',
        });

        await inputEl.trigger('blur');
        expect(wrapper.emitted()).toMatchInlineSnapshot(`
          {
            "blur": [
              [
                FocusEvent {
                  "isTrusted": false,
                },
              ],
            ],
            "eager": [
              [
                Event {
                  "isTrusted": false,
                },
              ],
            ],
            "input": [
              [
                Event {
                  "isTrusted": false,
                },
              ],
            ],
            "update:modelValue": [
              [
                "",
              ],
            ],
          }
        `);
        expect(wrapper.attributes().class).toContain('form-input--has-error');
    });

    it('renders valid state, when validate method is called and value passes validation check', async () => {
        const wrapper = factory({
            props: {
                modelValue: 'email@domain.sk',
            }
        });

        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        wrapper.vm.validate();

        await wrapper.vm.$nextTick();
        const stateEl: DOMWrapper<HTMLInputElement> = wrapper.find(STATE_SELECTOR);
        expect(stateEl.exists()).toBe(true);
        expect(wrapper.attributes().class).toContain('is-valid');
        expect(stateEl.html()).toMatchInlineSnapshot(`
          "<div class=\\"form-errors\\" data-test=\\"form-input__errors\\">
            <ul>
              <!--eslint-disable vue/no-v-html-->
            </ul>
          </div>"
        `);
    });

    it('renders invalid state, when validate method is called and value doesnt pass validation check', async () => {
        const wrapper = factory({
            props: {
                modelValue: 'email domain',
            }
        });

        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        wrapper.vm.validate();

        await wrapper.vm.$nextTick();
        const stateEl: DOMWrapper<HTMLInputElement> = wrapper.find(STATE_SELECTOR);
        expect(stateEl.exists()).toBe(true);
        expect(wrapper.attributes().class).toContain('has-error');
        expect(stateEl.html()).toMatchInlineSnapshot(`
          "<div class=\\"form-errors form-errors--is-error\\" data-test=\\"form-input__errors\\">
            <ul>
              <!--eslint-disable vue/no-v-html-->
              <li class=\\"form-errors__error\\">Zadajte správny e-mail.</li>
            </ul>
          </div>"
        `);
    });

    it('clears rendered error state and value', async () => {
        const wrapper = factory({
            props: {
                modelValue: 'email domain',
            },
        });

        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        wrapper.vm.validate();
        await wrapper.vm.$nextTick();

        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        wrapper.vm.clear();
        await wrapper.vm.$nextTick();

        const stateEl: DOMWrapper<HTMLInputElement> = wrapper.find(STATE_SELECTOR);
        const inputEl: DOMWrapper<HTMLInputElement> = wrapper.find(INPUT_SELECTOR);
        expect(stateEl.exists()).toBe(false);
        expect(inputEl.element.value).toBe('');
        expect(wrapper.attributes().class).not.toContain('has-error');
    });

    it('subscribes to a global handler on mount', async () => {
        const wrapper = factory();

        expect((wrapper.vm as any).onItemMounted).toBeCalled();
    });

    it('unsubscribes from a global handler on before unmount', async () => {
        const wrapper = factory();

        wrapper.unmount();

        expect((wrapper.vm as any).onItemBeforeUnmount).toBeCalled();
    });
});
