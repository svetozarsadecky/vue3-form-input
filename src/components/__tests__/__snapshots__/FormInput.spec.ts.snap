// Vitest Snapshot v1

exports[`FormInput props, slots & events > renders component 1`] = `
"<div class=\\"form-input\\">
  <div class=\\"form-input__field\\">
    <!--v-if--><label for=\\"full-email\\" class=\\"form-input__wrapper\\">
      <!--v-if--><input id=\\"full-email\\" accept=\\"\\" autocomplete=\\"username email\\" class=\\"form-input__input\\" name=\\"email\\" pattern=\\"[a-z0-9._%+-]+@[a-z0-9.-]+\\\\.[a-z]{2,}$\\" type=\\"email\\" data-test=\\"form-input__input\\"><span class=\\"form-input__label\\" data-test=\\"form-input__label\\">E-mail</span>
    </label>
    <!--v-if-->
    <!--v-if-->
  </div>
  <!--v-if-->
  <!--v-if-->
</div>"
`;
